### ISSUE TITLE
Question About Handling Pair Position Updates After Merges

### ISSUE THREAD

**Issue by depshad** [gautierdag/bpeasy#1]
**Description**
I've been going through the BPE implementation and noticed something that I'd like to get your thoughts on. After a pair is merged within a sentence, the code updates the positions and frequencies of the adjacent pairs. However, I’m curious about how the code handles the positions of all subsequent pairs in the sentence.

**Context**
For example, with the sequence [a, b, c, d, e, f]:

If the pair (b, c) is merged into bc, the sequence becomes [a, bc, d, e, f].
It seems like the code updates the adjacent pairs (a, bc) and (bc, d), but I’m wondering if the positions of the subsequent pairs (d, e) and (e, f) are recalculated to account for the shift caused by the merge.

**My Question**
Is there a mechanism in the code that ensures the positions and frequencies of all pairs in the sentence are recalculated after each merge? 

I’d appreciate any insights on whether this is the intended behavior or if there’s an existing solution in the code that addresses this.

**Comment by gautierdag** [comment]
Hey thanks for your question!

I think what you're looking for is the update to the `global_pair_count` and `global_pair_positions` which happens after propagating the changes.

https://github.com/gautierdag/bpeasy/blob/3b232e90d33de960be258cd6ce7107ef8d89676c/src/lib.rs#L314-L333

The merge operator will also be useful, specifically:

https://github.com/gautierdag/bpeasy/blob/3b232e90d33de960be258cd6ce7107ef8d89676c/src/lib.rs#L111-L117

Apologies, there are a lot of optimizations that make the comprehension a little overwhelming.

If you look at the code above, the `global_pair_positions` refers to the idxs of the sentences rather than the position of pairs within each sentence. 

Maybe in simpler terms:

1. I have a text input: `"hello hell"`. 
2. I use a regex that splits on space: `["hello", " ", "hell"]`
3. I then tokenize these into base UTF tokens `[[1, 2, 3, 3, 4], [5], [1, 2, 3, 3]]` (Note the integers I have chosen here are not the actual ones).

4.a) Each of these sequences is represented with linked list datastructure `Sentence`
4.b) Single length sentences get removed so that you have: `[[1, 2, 3, 3, 4], [1, 2, 3, 3]]`. (I just pushed an update in `0.1.3` to do this)

5. I count pairs: 
```
(1, 2): 2
(2, 3): 2
(3, 3): 2
(3, 4): 1
```
and find locations (set of idxs of each `Sentence` where the pair occurs at least once):
```
(1, 2): {0, 1}
(2, 3): {0, 1}
(3, 3): {0, 1}
(3, 4): {0}
```

6. Find the most common pair and for each sentence where it appears and call `.merge`

Do steps 5/6 recursively until done.

Hope that clears it up!
